distributed/scheduler.py

```diff
@@ -3034,46 +2851,46 @@ def decide_worker(dependencies, stacks, stack_duration, processing, who_has,
     workers = frequencies([w for dep in deps
                              for w in who_has[dep]])
     if not workers:
-        workers = stacks
+        workers = occupancy
     if valid_workers is not True:
         workers = valid_workers & set(workers)
         if not workers:
             workers = valid_workers
             if not workers:
                 if key in loose_restrictions:
-                    return decide_worker(dependencies, stacks, stack_duration,
-                            processing, who_has, has_what, True, set(), nbytes,
-                            ncores, key)
+                    return decide_worker(dependencies, occupancy, who_has,
+                            True, set(), key, objective=objective,
+                            nbytes=nbytes, has_what=has_what, ncores=ncores)
+
                 else:
                     return None
-    if not workers or not stacks:
+    if not workers or not occupancy:
         return None
 
     if len(workers) == 1:
         return first(workers)
 
-    # Select worker that will finish task first
-    def objective(w):
-        comm_bytes = sum([nbytes.get(k, 1000) for k in dependencies[key]
-                          if w not in who_has[k]])
-        stack_time = stack_duration[w] / ncores[w]
-        start_time = comm_bytes / BANDWIDTH + stack_time
-        return start_time
+    if objective is None:
+        def objective(w):
+            comm_bytes = sum([nbytes.get(k, 1000) for k in dependencies[key]
+                              if w not in who_has[k]])
+            stack_time = occupancy[w] / ncores[w]
+            start_time = comm_bytes / BANDWIDTH + stack_time
+            return (start_time, len(has_what[w]))
```

Comment:
This is disturbingly similar to the `worker_objective()` method, except that it uses `len(has_what[worker])` as a proxy for memory occupation. Is this deliberate?